### ISSUE TITLE
global abbreviations but buffer-local &iskeyword

### ISSUE THREAD

**Issue by Konfekt** [vim/vim#5074]
Whether `:ab` throws an error `E474` depends on its non-keyword characters (`:help abbreviations`). For example, if `@` is not a keyword character but `a` is, then `:ab @aa a` throws an error `E474`.

But `:ab` creates a global abbreviation, whereas `&iskeyword` is buffer-local. If a global abbreviation is defined in the first buffer, but this definition would throw an error `E474` in a second buffer with a different value of `&l:iskeyword`, then the abbreviation simply does not expand in the second buffer.

Therefore, it seems more consistent if the definition of a global abbreviation never threw an error (and simply does not expand in buffers whose value of `&l:iskeyword` does not permit so). Instead maybe a warning could be given if the abbreviation contains non-word characters (`\W`) (suppressible by `:silent ab`).

**Comment by Konfekt** [comment]
Regarding backwards-compatibility, changing the error message `E474` by a warning that under the current settings of `&l:iskeyword` the abbreviation (referring to the :help abbreviations`) does not expand seems the best way out.

**Comment by Houl** [comment]
I'd choose global-abbrev left-hand-sides that will work with a lot of local `'isk'` settings (certainly not the case for `@aa`).  And to avoid the error, you can set `'isk'` to a sane value (ie `isk&`) before defining the global abbrevs, and later restore it.

**Comment by Konfekt** [comment]
Yes, from a user's perspective it makes sense to define global abbreviations only for `&iskeyword` being equal to the global default `set iskeyword&` which happens to be a word character `\w`. 

Since the error message can easily be circumvented by intermediately changing `&iskeyword`, ideally the error message for global abbreviations should become a warning whenever the abbreviation is invalid for the global default `set iskeyword&`; 
that is, whenever the non-word characters (`\W`) in the alias name do not enclose the word characters Or are not all last Or are not all first and followed by at most one word character.

**Comment by brammool** [comment]
I think changing the behavior doesn't help.  We do need the error, otherwise users make mistakes and add abbreviations that will never work.  Also, it makes sense the the moment an abbreviation is defined the current 'iskeyword' value is used.  So the user doesn't get a warning when using the abbreviation in a buffer where 'iskeyword' has a different value, but is that really a problem?

**Comment by Konfekt** [comment]
> So the user doesn't get a warning when using the abbreviation in a buffer where 'iskeyword' has a different value, but is that really a problem?

That's a smaller different problem.

> We do need the error, otherwise users make mistakes and add abbreviations that will never work.

Yes, the user better be notified, but it would be more consistent if the global abbreviation would still be created: Impeding the user to define a global abbreviation that could work in all buffers but the current one (due to the  buffer-local value `&l:iskeyword`) is undue (and an intermediary change of `&l:iskeyword` works around it).
Seeing error `E447` after defining a global abbreviation gives the user the impression that this abbreviation is globally invalid, but in fact it is only locally. Thus, the user should be warned that the abbreviation does not expand in the current buffer, or at least the error message could reflect that.
